### DISCUSSION TITLE
Silence when using custom pipewire configuration with easyeffects. Suggestions?

### DISCUSSION THREAD

**Discussion by Ninjastorm380** [wwmm/easyeffects#2081] [Q&A] [answered]
Hello! After a recent commit, easyeffects stopped working properly as a default device (please don't rail on me, am one of those weird edge cases that needed this -..- ), so I reconfigured easyeffects as per the instructions in [here](https://github.com/wwmm/easyeffects/discussions/1836). Now, I have complete silence when pipewire is used with this configuration file: [pipewire.conf.txt](https://github.com/wwmm/easyeffects/files/10468028/pipewire.conf.txt). since it seems that my config is to blame here, I was wondering if anyone could proofread my handiwork?

Side note: config is used for isolating and capturing audio streams for streaming to twitch.

**Comment by wwmm** [comment, marked as answer]
Run the command `pw-dot` hile EasyEffects is running and an audio player too. The output file can be viewed with the command `xdot` and will show how filters and devices are linked.

I do not remember any recent commit that is even remotely close to device management. My guess is that the custom configuration is somehow breaking our custom routing. pw-dot output should help to verify if this is the case.

**Reply by Ninjastorm380** [reply]
After running these commands and rebooting a couple of times (compiling/installing xanmod kernel stuffs) the issue has magically disappeared?!? I'm going to keep using this setup over the next month, maybe this is a intermittent bug?
here's what helvum sees while it's working, btw: ![image](https://user-images.githubusercontent.com/40574927/213771588-13c96e7b-17a8-43bc-8c1b-02d9616e4c0d.png)
firefox nightly is playing some music from newgrounds for testing.

**Reply by wwmm** [reply]
>maybe this is a intermittent bug?

The audio routing in the image seems fine. So maybe PipeWire got confused while trying to activate its streams state restoration function and messed something. Strange indeed...

**Reply by Ninjastorm380** [reply]
@wwmm Issue has popped up again. `pw-dot` returns nothing, so I cannot use `xdot` to view anything. helvum graph for non working session:
![image](https://user-images.githubusercontent.com/40574927/213785719-a98aa28a-1444-48c3-9ed7-a47fb2b2c9c4.png)

**Reply by Ninjastorm380** [reply]
~~I'm looking at this, and it definitely seems like firefox didn't get routed at all.~~ Scratch that, I'm blind

**Reply by wwmm** [reply]
>pw-dot returns nothing, so I cannot use xdot to view anything. 

This is unusual. It should produce a file named `pw.dot` or similar in the directory where the command was executed. The advatange of using pw-dot is that it also shows the links state(active,paused, etc).

**Reply by wwmm** [reply]
Is `pw-top` showing many error counts in one of its columns?

**Reply by Ninjastorm380** [reply]
there are definitely errors
![image](https://user-images.githubusercontent.com/40574927/213789652-be49c798-cb88-4ca2-99e2-57d5f91e7767.png)

**Reply by Ninjastorm380** [reply]
I probably should mention mention that I'm using `wireplumber` for the session manager, as pacman states that `pipewire-media-session` is depreciated and to expect it to be removed. `wireplumber` is also default on endeavouros arch.

**Reply by wwmm** [reply]
I have seen a large amount of errors in other situations too. It was not like this in the past. I think since one or two PipeWire releases ago xrun errors became more frequent. We may be facing some kind of Pipewire regression. The changes in EasyEffects in the last months have nothing to do with the low level filter management.

**Reply by Ninjastorm380** [reply]
@wwmm I managed to get `pw-dot` to provide an output, but I cannot upload it without mangling the extension, and the `xdot` viewer shows it as way too small

**Reply by Ninjastorm380** [reply]
here's the pw.dot file [pw.dot.txt](https://github.com/wwmm/easyeffects/files/10469703/pw.dot.txt)

**Reply by Ninjastorm380** [reply]
for giggles, I installed the now depreciated `pipewire-media-session`, and I have sound back. possible regression in `wireplumber` then?

**Reply by wwmm** [reply]
>here's the pw.dot file

It looks fine. So what was within EasyEffects reach was done. The problem must be deeper in Pipewire/Wireplumber. 

>I installed the now depreciated pipewire-media-session, and I have sound back. possible regression in wireplumber then?

EasyEffects relies on the media session manager to link streams to our virtual devices. So it may be something in wireplumber. I haven't used `pipewire-media-session` in many months as wireplumber is supposed to replace it in the future.

**Reply by Ninjastorm380** [reply]
perhaps you could experiment with my config and figure out what breaks? am still relatively new to kenerl dev stuff, so this is already an over my head problem >..<

**Reply by Ninjastorm380** [reply]
@wwmm for now, I have switched to git AUR package `pipewire-media-session-git`. this seems to solve the issue for now

**Reply by wwmm** [reply]
>perhaps you could experiment with my config and figure out what breaks?

The only difference between our configuration files that seem relevant is that I do not have any loopback device. I don't really have an use for them. But based on pw-dot and Helvum output the links are done and active. So at first sight it does not seem loopback devices are the trigger. On the other hand loopbacks are made in a different way than EasyEffects filters. They are created with streams and streams will be touched by WirePlumber.

When there is no audio all apps in Helvum image are affected or only the ones in the loopback?

**Reply by wwmm** [reply]
>for now, I have switched to git AUR package pipewire-media-session-git. this seems to solve the issue for now

Usually when something wrong happens when using WirePlumber but it doesn't when using the built-in media session manager it is a good idea to report the problem to WirePlumber developers. In most cases I have seen so far when such a thing happens it is a flaw in WirePlumber. And as WirePlumber is supposed to replace the built-in manager in the future it is better to try to fix its problems now.

**Reply by Ninjastorm380** [reply]
@wwmm everything is silent while using wireplumber, not just applications sending to virtual streams.
I'll go report this to the wireplumber devs, then.

**Reply by pinkflames** [reply]
@wwmm : slight correction, that *future* arrived more than a year ago. :wink:

The legacy pipewire-media-session is and has been truly deprecated for a long time now. Whether or not this is a regression or design inevitability is sadly something I can't say. @Ninjastorm380 will have to wait and see what WP developers have to say in https://gitlab.freedesktop.org/pipewire/wireplumber/-/issues/409 .

**Reply by wwmm** [reply]
>slight correction, that future arrived more than a year ago. wink

Oh :smile: . Thanks. I had no idea the official move to WirePlumber had already been made.

**Reply by wwmm** [reply]
@Ninjastorm380 which WirePlumber version is available to you? I wonder if what you are seeing is a bug supposed to be fixed in its latest release 

https://gitlab.freedesktop.org/pipewire/wireplumber/-/merge_requests/450
https://github.com/Audio4Linux/JDSP4Linux/issues/74#issuecomment-1301701043

**Reply by Ninjastorm380** [reply]
@wwmm arch can follow master with `wireplumber-git`

**Reply by Ninjastorm380** [reply]
that said, latest non git is 4.13-1

**Reply by wwmm** [reply]
Ok. I forgot you are also on Arch Linux.

**Reply by Ninjastorm380** [reply]
@wwmm just switched to aforementioned git package, and seems to be functional. will keep testing to prove that.

**Reply by Ninjastorm380** [reply]
@wwmm `wireplumber-git` is now confirmed to fix the previously mentioned difficulties. I wonder what changed between stable and git?

**Reply by wwmm** [reply]
>I wonder what changed between stable and git?


I have no idea. At least in Wireplumber's commits names since the last release nothing rings a bell.
